vault/router.go

```diff
@@ -802,20 +812,117 @@ func (r *Router) LoginPath(ctx context.Context, path string) bool {
 	remain := strings.TrimPrefix(adjustedPath, mount)
 
 	// Check the loginPaths of this backend
-	loginPaths := re.loginPaths.Load().(*radix.Tree)
-	match, raw, ok := loginPaths.LongestPrefix(remain)
-	if !ok {
+	pe := re.loginPaths.Load().(*loginPathsEntry)
+	match, raw, ok := pe.paths.LongestPrefix(remain)
+	if !ok && len(pe.wildcardPaths) == 0 {
+		// no match found
 		return false
 	}
-	prefixMatch := raw.(bool)
 
-	// Handle the prefix match case
-	if prefixMatch {
-		return strings.HasPrefix(remain, match)
+	// Matching Priority
+	//  1. prefix
+	//  2. exact
+	//  3. wildcard
+	if ok {
+		prefixMatch := raw.(bool)
+		if prefixMatch {
+			// Handle the prefix match case
+			return strings.HasPrefix(remain, match)
+		}
+		if match == remain {
+			// Handle the exact match case
+			return true
+		}
 	}
 
-	// Handle the exact match case
-	return match == remain
+	// check Login Paths containing wildcards
+	reqPathParts := strings.Split(remain, "/")
+	for wcPath := range pe.wildcardPaths {
+		if wcPath == "" {
+			continue
+		}
+		isPrefix := false
+
+		currWCPath := wcPath
+		if currWCPath[len(currWCPath)-1] == '*' {
+			isPrefix = true
+			currWCPath = currWCPath[0 : len(currWCPath)-1]
+		}
+
+		splitCurrWCPath := strings.Split(currWCPath, "/")
+
+		if len(reqPathParts) < len(splitCurrWCPath) {
+			// check if the path coming in is shorter; if so it can't match
+			continue
+		}
+		if !isPrefix && len(splitCurrWCPath) != len(reqPathParts) {
+			// If it's not a prefix we expect the same number of segments
+			continue
+		}
+
+		segments := make([]string, 0, len(splitCurrWCPath))
+		matchFound := true
+		for i, wcPathPart := range splitCurrWCPath {
+			switch {
+			case wcPathPart == "+", wcPathPart == reqPathParts[i]:
+				segments = append(segments, reqPathParts[i])
+			case isPrefix && i == len(splitCurrWCPath)-1 && strings.HasPrefix(reqPathParts[i], wcPathPart):
+				segments = append(segments, reqPathParts[i:]...)
+			default:
+				// we encounted segments that did not match
+				// this flag will prevent false positives when currWCPath is a prefix
+				matchFound = false
```

Comment:
Is `matchFound` needed here? It looks like trim the prefix character by doing `currWCPath = currWCPath[0 : len(currWCPath)-1]` and we're building up segments to be concatenated back to a full string during this process.